### PR TITLE
Restructures Hulk as a job-bannable role.

### PR DESCRIPTION
**Pull request by DreamwalkerSisyphe**
:cl: The Dreamweaver
admin: Gives admins the ability to role-ban a user from utilizing the Hulk trait.
/:cl:
 
This change has already been approved by at least one headmin.

I have gotten further words from headmin that a hulk ban would be given as a full punishment and that those banned do not deserve exceptions during antag rounds: https://imgur.com/a/kweHS0N

### CODE REVIEW

**Comment by PKPenguin321** [comment]
interesting idea

**Comment by yoyobatty** [comment]
Considering people proceed to delete every single wall on the station once this is obtained, I approve.

**Comment by TheDreamweaver** [comment]
https://v637g.app.goo.gl/CJWBVUT7wzNzjoV98

**Comment by Ratvar** [comment]
Hulk whitelist when

File: code/datums/mutations/hulk.dm
Comment (PKPenguin321) [reply code/datums/mutations/hulk.dm]: Maybe just put a big red "You are banned from hulk." message or something

Comment (TheDreamweaver) [reply code/datums/mutations/hulk.dm]: It plays into the lore that NT is always one step ahead of the crew just like the US military :). I guess I can make it more something along the lines of the "gods" cursing you, since admins=gods.

**Comment by Jordie0608** [comment]
Add this to the banning panel per #40039

**Comment by TheDreamweaver** [comment]
@jordie this PR already adds it to the role banning panel. Also your pr that you referenced is unnecessary given the PR I just pushed out previous to this one, which adds those two potions to the same role ban panel.

**Comment by TheDreamweaver** [comment]
Nvm, I'm actually insane and just ended up doing it from my phone using the web editor. I want to die now.

File: code/datums/mutations/hulk.dm
Comment (coiax) [inline code/datums/mutations/hulk.dm]: This seems a little OOC. Can't we make up something about Centcom putting in genetic kill switches in some of their employees... genes to stop them abusing hulk? Or something?

Comment (JJRcop) [reply code/datums/mutations/hulk.dm]: I've [argued against this earlier](https://github.com/tgstation/tgstation/pull/40037#discussion_r214502175). In an earlier commit, this was exactly what was said here. NT directly restricting people from hulk when hulk has to be researched and discovered by geneticists, doesn't make a 
lot of sense. I encourage you to come up wih something better than an NT cop-out.

Comment (coiax) [reply code/datums/mutations/hulk.dm]: I don't have any better ideas, but I thought we were aiming for less OOC in IC statements like SPACEGODS ARE ASHAMED OF YOU

Comment (octareenroon91) [reply code/datums/mutations/hulk.dm]: "On second thought, your muscles feel just fine :)"

Comment (PKPenguin321) [reply code/datums/mutations/hulk.dm]: Again, it might be better to not even dance around it and just say "You are banned from being a hulk."

Comment (TheDreamweaver) [reply code/datums/mutations/hulk.dm]: I'm leaving it. The other in game role bans are not icly subtle not are they completely oocly blunt. If the maintainers have a huge problem with it then we'll talk, but it has very little impact on anything itself.

**Comment by Shdorsh** [comment]
Gone are the days where you'd just take an e-cigarette, fill it up with around half half unstable mutagen and charcoal, then go on an autistic rampage before being permastunned by seizures and turning invisible with no one to help you anymore, I guess.

**Comment by Jordie0608** [comment]
This will need an override for wizards since it would interfere with one who chose the mutate spell while banned from hulk.

**Comment by Iamgoofball** [comment]
Why don't we just remove hulk and antag lock it

**Comment by TheDreamweaver** [comment]
@jordie0608 why should someone who couldn't even use hulk appropriately in a normal setting be allowed an opportunity to murderbone with it?

**Comment by JJRcop** [comment]
I think that instead of removing people from hulk if they're hulkbanned, and checking for it when it's acquired, you should check for it when the hulk tries to do any hulk punches, and turn them into normal punches. Then you can change the current message to "The gods look down upon you... Your large green muscles feel just as weak as before.". Then, being hulkbanned won't save you from the downsides of being hulk, like non-human-ness.

**Comment by TheDreamweaver** [comment]
@jjrcop That's convuluted and needless. Plus they would still be able to reap the benefits of the antistuns. Ideally if someone was banned from using hulk, it's because they seriously abused it, and therefore shouldn't be reaping any benefits. I'm not going to go through and have a check to negate every aspect of hulk just they can be counted as valids. People banned from traitor don't still spawn as traitors but without a uplink and permission to kill.

**Comment by octareenroon91** [comment]
>>wizard's mutate spell

>why should someone who couldn't even use hulk appropriately in a normal setting be allowed an opportunity to murderbone with it?

##### _maybe it should be rolled back for any antag type_

More directly, I'm pretty sure a few of us guessed the bans would come for excess vandalism and crime as crew.

**Comment by TheDreamweaver** [comment]
@octareenroon91 I'm confused by what you mean by "rolled back for any antag type". If your suggesting making it so any antag can bypass a hulk ban, then I'll repeat my previous answer. Those who messed up so badly as to get banned from using hulk should fully reap the consequences, including not being able to utilize it for antag roles. If that's not what you're suggesting, then please clarify further.

**Comment by TheDreamweaver** [comment]
I am willing to possibly make it so wizards who are hulk banned are blocked from even purchasing the mutation spell as to avoid accidentally purchasing a spell that you forgot you couldn't use, but perhaps a wizard might still want utilize the laser vision, even if they can't use hulk.

**Comment by octareenroon91** [comment]
We're deep in hypotheticals now, arguing why someone would get a hulk ban.

It's getting to the shallow side of the separation between code and server, but maybe we should seek some admin insight.

**Comment by PKPenguin321** [comment]
The mutate spell is a different thing from hulk even if it uses it codewise imo

**Comment by TheDreamweaver** [comment]
@octareenroon91 I agree that this is an adminstration issue and the reason for banning is important to the context.

**Comment by TheDreamweaver** [comment]
@pkpenguin321 You still get hulk, regardless. The only thing that's different is the context you received it which only matters in context of why you got the ban. That just means admins need to be aware of the consequences of hulk banning someone.

**Comment by TheDreamweaver** [comment]
Alright talked to headmin, and got my answer. Update in the description.

**Comment by Iamgoofball** [comment]
What? No, stop listening to idiots who have no say over the codebase @TheDreamweaver 

@optimumtact can you clarify for this guy what separation of codebase and server means

**Comment by Iamgoofball** [comment]
Additionally, this shit can now be used in a roundabout way to metagame if someone is really X or Y.

**Comment by Iamgoofball** [comment]
This is not an administration issue at all. This is the admins sticking their noses into a gameplay problem of hulk being fucking garbage.

File: code/datums/mutations/hulk.dm
Comment (Iamgoofball) [inline code/datums/mutations/hulk.dm]: Matter of fact, what is this hardcoded shit?

Go home, and write a proper genetics jobban for this. Don't half-ass this.

File: code/modules/admin/DB_ban/functions.dm
Comment (Iamgoofball) [inline code/modules/admin/DB_ban/functions.dm]: GLOB this list.

File: code/modules/admin/topic.dm
Comment (Iamgoofball) [inline code/modules/admin/topic.dm]: This is copypasted. Undo the copypaste here and clean up this proc.

File: code/modules/admin/topic.dm
Comment (Iamgoofball) [inline code/modules/admin/topic.dm]: Why are these lists hardcoded?

File: code/datums/mutations/hulk.dm
Comment (Iamgoofball) [inline code/datums/mutations/hulk.dm]: This doesn't seem like the right proc to call. Did you even test this?

**Review by Iamgoofball — CHANGES_REQUESTED** [review]

**Comment by TheDreamweaver** [comment]
@iamgoofball I'd appreciate it if you didn't needlessly insult people just because don't agree with them. And this is absolutely an adminstration issue as context is super important to this PR, and that context is for what reasons admins would be using this ban and why. Your comment is a prime example of why to much separation is a bad thing.

**Comment by TheDreamweaver** [comment]
Also the changes you are proposing would require full DB changes, which is not what this PR should be worrying about.

**Comment by Iamgoofball** [comment]
There is full separation.

The admins have no say over gameplay decisions like this. They are adding a ban to a feature because of concerns of misuse and overuse, when instead we should just be removing it.

You are cluttering the codebase with hacky garbage you do not seem to have tested in order to fufill some random headmin's wishes who **is about to be replaced in the headmin elections anyways**.

Instead of doing the proper thing, which is to remove the feature that serves no purpose than to nullify the main combat method(stuns) and easy wall destruction.

Hulk has no place in this codebase. Tell your headmin to fucking get in here and explain exactly why they need this ban so bad and why it's better than just removing the feature we've deemed broken enough to not allow even antags to use.

**Comment by Iamgoofball** [comment]
>Also the changes you are proposing would require full DB changes

You very clearly didn't read the goddamn review then. Use your eyes and your critical thinking skills.

**Comment by TheDreamweaver** [comment]
@iamgoofball I constructed this PR is the way I was instructed to by the involved maintainers. I was also told to hold off one reworks like the ones you have proposed until a separate PR.

**Comment by Iamgoofball** [comment]
Yeah, I don't believe you. Go clean that code up. No maintainer would tell you to copypaste.

**Comment by Iamgoofball** [comment]
Can a gitban be placed on @TheDreamweaver for abusing a github bug to resolve reviews?
[img](https://file.house/JeHr.png)

**Comment by TheDreamweaver** [comment]
Sorry goof, what you're proposing is agreed upon to not be within the span of what this PR is supposed to me
Unfortunately I'm away from home and therefore not in a position to have long conversations about this. If you'd like to have further conversation about this when I get back late Monday, is be willing to do that. As for now, I'd appreciate it if you didn't clutter this PR.

File: code/modules/admin/topic.dm
Comment (TheDreamweaver) [reply code/modules/admin/topic.dm]: Refer to above comment

File: code/modules/admin/topic.dm
Comment (TheDreamweaver) [reply code/modules/admin/topic.dm]: Refer to above comment

**Comment by Iatots** [comment]
1) How can someone without enough impulse control to not repeatedly be a nuisance when hulked to the point of receiving a hulk ban be trusted to not repeat his disruptive behavior?

2) The "tough on crime" schtick of hulkbanning antagonists is the ultimate irony. You are making a concession, you are letting rampaging shitters off the hook from a stronger sentence by removing one avenue out of many at their disposals. What's next? Hands too weak to turn a tank valve? Hands too sweaty to hold a banana or bar of soap? Forced clumsy trait? Is the staff at the helm so burned out that they can only find solace in importing a population of drooling retards and abusing their lobotomized meatsacks with """IC conflicts"""?

**Comment by TheDreamweaver** [comment]
As far as has been explained to me by the head in, this would only be issued to handle situations that would be justified in a normal ban, so if anything this is giving said players a second chance to prove themselves via means they wouldn't have if full banned. Assuming no we hold admins responsible and get it in writing, this will not be used to police how players play the game anymore than a normal ban would. The coder/admin interaction should involve providing them with helpful tools to aide them in their duties, not to try to police admins via code.

**Comment by obscolene** [comment]
@Nabski89 scared?

**Comment by Nabski89** [comment]
Nope, because I hulk responsibly and other than the one time I've punched a syndicate bomb to "Death" I've never had an admin say something to me about my hulk usage.

The trick is to keep smashing random small shit like tables and corners so you're constantly GRAHHHHing on the radio, but not actually annoying or fighting people.

**Comment by TheDreamweaver** [comment]
I don't think I have to explain why a geneticist ban wouldn't solve the problem. And again, if someone is still confused about why or how this would be used, I linked a response from a headmins in the description. I don't know why people are trying to admin-police in thread. As stated above, the coder/admin interaction involves providing admins with useful tools so they can better do their job. Headmins/admins have already stated to me that they would find this useful. If for some reason you do not trust your admins to do their job, it is not my job as a coder to police them, as unlike players, admins can be much more easilly held responsible. If you have any further questions about how this would be used and **not** about the code itself, please ask the headmins/admins themselves.

**Comment by TheDreamweaver** [comment]
If you want to get an idea of how this would work administratively, just take a look at Mind Transfer Potions bans, which by no coincidence, are implemented almost exactly the same. Any further administrative  questions may be taken to the administrators themselves.

**Comment by DaxDupont** [comment]
We need to replace hulk really they are hilariously unbalanced and is to easy to get

**Comment by TheDreamweaver** [comment]
@daxdupont This sentimate may be true, but that's not within the span of this PR. Feel free to accomplish this in a PR of your own, heck, I'd even be willing to help.

**Comment by Kagus** [comment]
I mean, it's most likely not going to hurt anything to have a hulk-specific ban. I've seen multiple players who act semi-responsibly during regular play behave themselves all the way up until the point when someone gives them hulk, or they manage to get it for themselves. At that point, it's a game of "The walls are valids".

Just because there aren't many logged notes/bans specifically regarding hulk behavior doesn't mean that there won't be, or that the behavior doesn't exist. I suppose a somewhat better metric would be to implement the ban ability for a while, see if it gets used, and see what the response is from community/mins after such usage.

Also, phonegramming... You mad bastard.

**Comment by ExcessiveUseOfCobblestone** [comment]
What prompted this? It doesn't make a lot of sense.

I'm MCDONALDS^3 and I think this is too specific, especially when it explicitly gets in the way of potential antag gimmicks.

**Comment by TheDreamweaver** [comment]
@excessiveuseofcobblestone I was told it might be useful by certain headmin, so I coded it, simple as that. Kinda like the mind transfer/sentience potion ban that were implemented just before this in the exact same way that no one spoke up about because unlike hulk, it doesn't mess with their meta.

**Comment by TheDreamweaver** [comment]
Also again, I'm taking the word of headmins here, so please, if you have any further questions about how this would be utilized, and not the code themselves, please go ask them first as to avoid cluttering this page with further comments/questions that I have either already answered or can not answer any further myself. I'm not currently an admin so I can not answer past the info I've received from them.

**Comment by ExcessiveUseOfCobblestone** [comment]
The last PR was okay if only because it separated very-much-needed separated roles because banning them from one equaled banning them from all (sad 😢). I'm perfectly okay with banning people for sentience roles since that is a role.

Banning people from using certain IC features, even while antag, is foolish (hulk OR mindtransfer). If you are going to let them play antag then give them the opportunity to utilize every aspect of the game to create an interesting experience since that's their job as antagonist. Otherwise they should be outright gamebanned. I'm genuinely amazed at the people who tout this while also touting not wanting TDM SS13. You only encourage people to just ebow/esword.

PSA: I'll unban anyone who asks for it during their antag round if this gets merged regardless of posts below. Same for mind-transfer now. Consider it a plus if I forget to readd the ban.

**Comment by ExcessiveUseOfCobblestone** [comment]
Also if you titled the PR appropriately then we would have been having the conversation their instead of here. It has nothing to do with my preferential murderbone method or whathave you.

**Comment by TheDreamweaver** [comment]
@excessiveuseofcobblestone This seems like a serious adminstrative issue if admins are not in agreement with their headmins to the point they're repealling bans, or not enforcing rules the same. I'm not here to make adminstrative decisions, that's not my job as a coder. Sure I might get frustrated as I put a bit of time in creating this after being told it might be useful, but at the end of the day this PR getting merged affects the adminstration, not me. It seems like maybe an admistrative discussion needs to be had (which is why I've been encouraging people to go talk to their headmins instead of screaming into the void here).

Also my title and description make I perfectly clear what this PR is about, so not sure what you are implying.

**Comment by TheDreamweaver** [comment]
@excessiveuseofcobblestone I can see the confusion, but the reason why it's phrased that way is because, unlikethis hulk PR, mind transfer potions have always been bannable, it's just that they relied on the xenomorhp role ban. Hence it being a fix of the mentioned issue, not a complete new addition.

**Comment by ExcessiveUseOfCobblestone** [comment]
I don't think anyone would be banned for using mind control anyways, it's never gotten to that point from my experience. People (myself included) did not know that was a thing, or else we'd be pitchforks and torches there.

In your defense of this PR it's not your fault that the headmin did not speak to the people meant to utilize this. We all make mistakes, headmins included.

**Comment by TheDreamweaver** [comment]
@ExcessiveUseOfCobblestone That's fair enough. Do you mind doing me a favor then? Since I don't currently hold an admin position, but you do, do you mind getting this sorted out with them so I can move on to other things? If so, I'd greatly appreciate it. Also I think this situation is a great example of why the coder/admin separation can be a bad thing.

**Comment by ninjanomnom** [comment]
I'm not a fan of the precedent this sets. Overly granular restrictions on players for an administration reason that mess with game balance isn't something we should be supporting.

**Comment by TheDreamweaver** [comment]
@ninjanomnom Understandable, though I'm surprised mind transfer potions bans were ever added then following the logic given.

**Comment by ExcessiveUseOfCobblestone** [comment]
They should be removed.

**Comment by ninjanomnom** [comment]
Maintainers aren't a hive mind and our opinions change like normal people. Plenty of things also get by without comment or much thought if it seems harmless.

**Comment by TheDreamweaver** [comment]
@ninjanomnom Ya I get that, we're all humans. It just makes things a bit confusing on my end.

**Comment by TheDreamweaver** [comment]
@Jordie0608 You should reopen the corresponding PR, since those changes will now have not been made.

**Comment by ExcessiveUseOfCobblestone** [comment]
yes in terms of being able to be "job"-bannable.

no in terms of the feature itself.

**Pull request closed without merging**
